test/test_response.py

```diff
@@ -406,11 +407,96 @@ def close(self):
         self.assertEqual(next(stream), b'o')
         self.assertRaises(StopIteration, next, stream)
 
+    def test_mock_transfer_encoding_chunked(self):
+        stream = ["fo", "o", "bar"]
+        fp = MockChunkedEncodingResponse(stream)
+        r = HLHTTPResponse(MockSock)
+        r.fp = fp
+        resp = HTTPResponse(r, preload_content=False, headers={'transfer-encoding': 'chunked'})
+
+        i = 0
+        for c in resp:
+            self.assertEqual(c, stream[i])
+            i += 1
+
+    def test_mock_transfer_encoding_chunked_custom_read(self):
+        stream = ["foooo", "bbbbaaaaar"]
+        fp = MockChunkedEncodingResponse(stream)
+        r = HLHTTPResponse(MockSock)
+        r.fp = fp
+        r.chunked = True
+        r.chunk_left = None
+        resp = HTTPResponse(r, preload_content=False, headers={'transfer-encoding': 'chunked'})
+        expected_response = ['fo', 'oo', 'o', 'bb', 'bb', 'aa', 'aa', 'ar']
+        response = list(resp.read_chunked(2))
+        self.assertListEqual(expected_response, response)
+
     def test_get_case_insensitive_headers(self):
         headers = {'host': 'example.com'}
         r = HTTPResponse(headers=headers)
         self.assertEqual(r.headers.get('host'), 'example.com')
         self.assertEqual(r.headers.get('Host'), 'example.com')
 
+
+class MockChunkedEncodingResponse(object):
+
+    def __init__(self, content):
+        """
+        content: collection of str, each str is a chunk in response
+        """
+        self.content = content
+        self.index = 0
+        self.closed = False
+        self.exhausted = False
+        self.cur_chunk = ''
+
+    def get_chunk(self):
+        #if self.exhausted:
+        #    self.cur_chunk = ''
+        if len(self.cur_chunk) <= 0:
+            try:
+                chunk = self.content[self.index]
+            except IndexError:
+                self.exhausted = True
+                chunk = ''
+            else:
+                self.index += 1
+            self.cur_chunk = '%X\r\n%s\r\n'%(len(chunk), chunk)
```

Comment:
This won't work when you eventually convert the chunks to bytes, you'll have to use a different approach.